usr.bin/mdo/mdo.c

```diff
@@ -19,52 +20,277 @@
+		setcred_flags |= SETCREDF_RGID;
+	}
+	if (svgid_str) {
+		const char *errp = NULL;
+		wcred.sc_svuid = strtonum(svgid_str, 0, GID_MAX, &errp);
+		if (errp)
+			err(EXIT_FAILURE, "-Q: invalid GID");
+		setcred_flags |= SETCREDF_SVGID;
+	}
+
+	if (primary_group) {
+		struct group *gr = getgrnam(primary_group);
+		if (gr)
+			gid = gr->gr_gid;
+		else {
+			const char *errp = NULL;
+			gid = strtonum(primary_group, 0, GID_MAX, &errp);
+			if (errp != NULL)
+				err(EXIT_FAILURE, "invalid group '%s'", primary_group);
+		}
+		override_gid = true;
+	} else if (!uidonly && pw) {
+		gid = pw->pw_gid;
+		override_gid = true;
+	}
+
+	if (override_gid) {
+		wcred.sc_gid = wcred.sc_rgid = wcred.sc_svgid = gid;
+		setcred_flags |= SETCREDF_GID | SETCREDF_RGID | SETCREDF_SVGID;
+	}
+
+	if (supp_groups_str) {
+		char *copy = strdup(supp_groups_str);
+		if (!copy)
+			err(EXIT_FAILURE, "strdup failed");
+		char *tok = strtok(copy, ",");
+		while (tok) {
+			struct group *gr = getgrnam(tok);
+			gid_t g;
+			if (gr)
+				g = gr->gr_gid;
+			else {
+				const char *errp = NULL;
+				g = strtonum(tok, 0, GID_MAX, &errp);
+				if (errp != NULL)
+					err(EXIT_FAILURE, "invalid group '%s", tok);
+			}
+			supp_add = realloc(supp_add, sizeof(gid_t) * (add_count + 1));
+			if (!supp_add)
+				err(EXIT_FAILURE, "realloc failed");
+			supp_add[add_count++] = g;
+			tok = strtok(NULL, ",");
+		}
+		free(copy);
+		supp_reset = true;
+	}
+
+	if (group_mod_str) {
+		char *s = strdup(group_mod_str);
+		if (!s)
+			err(EXIT_FAILURE, "strdup failed");
+		char *tok = strtok(s, ",");
+		while (tok) {
+			if (strcmp(tok, "@") == 0) {
+				supp_reset = true;
+			} else if (tok[0] == '+' || tok[0] == '-') {
+				bool is_add = tok[0] == '+';
+				const char *gstr = tok + 1;
+				struct group *gr = getgrnam(gstr);
+				if (gr)
+					gid = gr->gr_gid;
+				else {
+					const char *errp = NULL;
+					gid = strtonum(gstr, 0, GID_MAX, &errp);
+					if (errp != NULL)
+						err(EXIT_FAILURE, "invalid group '%s'", gstr);
+				}
+				if (is_add) {
+					supp_add = realloc(supp_add, sizeof(gid_t) * (add_count + 1));
+					if (!supp_add) err(EXIT_FAILURE, "realloc failed");
+					supp_add[add_count++] = gid;
+				} else {
+					supp_rem = realloc(supp_rem, sizeof(gid_t) * (rem_count + 1));
+					if (!supp_rem) err(EXIT_FAILURE, "realloc failed");
+					supp_rem[rem_count++] = gid;
+				}
+			} else {
+				errx(EXIT_FAILURE, "invalid -s entry '%s'", tok);
+			}
+			tok = strtok(NULL, ",");
+		}
+		free(s);
+	}
+
+	if (supp_reset) {
+		gid_t *final = NULL;
+		size_t final_count = 0;
+
+		if (add_count > 0) {
+			final = malloc(sizeof(gid_t) * add_count);
+			if (!final)
+				err(EXIT_FAILURE, "malloc failed");
+
+			for (size_t i = 0; i < add_count; ++i) {
+				final[final_count++] = supp_add[i];
+			}
+		}
+
+		wcred.sc_supp_groups = final;
+		wcred.sc_supp_groups_nb = final_count;
+		setcred_flags |= SETCREDF_SUPP_GROUPS;
+	} else if (!supp_reset && (group_mod_str || !uidonly)) {
```

Comment:
`supp_reset` is necessary false here.

```suggestion
	} else if (group_mod_str || !uidonly) {
```